workers/data_refinery_workers/processors/salmon.py

```diff
@@ -285,33 +252,113 @@ def _run_salmon(job_context: Dict, skip_processed=SKIP_PROCESSED) -> Dict:
+def _run_salmontools(job_context: Dict, skip_processed=SKIP_PROCESSED) -> Dict:
+    """ Run Salmontools to extract unmapped genes. """
 
-    """
+    logger.debug("Running SalmonTools ...")
+    skip = False
+    unmapped_filename = job_context['output_directory'] + 'aux_info/unmapped_names.txt'
+    if skip_processed and os.path.exists(unmapped_filename):
+        logger.info("Skipping pre-processed SalmonTools run!")
+        skip = True
 
-    try:
-        tx_out = ro.r['::']('tximport', 'tximport')(job_context['output_directory'] + 'quant.sf', type="salmon", txOut=True)
-    except RRuntimeError as e:
-        # XXX/TODO: This doesn't work yet.
-        # But, eventually, it will look something like this. ^^
+    job_context['time_start'] = timezone.now()
+    if skip:  # If this procedure should be skipped, return immediately
+        job_context['time_end'] = timezone.now()
         return job_context
 
+    command_str = "salmontools extract-unmapped -u {unmapped_file} -o {output} "
+    output_prefix = job_context["output_directory"] + "unmapped_by_salmon"
+    command_str = command_str.format(unmapped_file=unmapped_filename,
+                                         output=output_prefix)
+    if "input_file_path_2" in job_context:
+        command_str += "-1 {input_1} -2 {input_2}"
+        command_str = command_str.format(input_1=job_context["input_file_path"],
+                                         input_2=job_context["input_file_path_2"])
+    else:
+        command_str += "-r {input_1}"
+        command_str= command_str.format(input_1=job_context["input_file_path"])
+
+    logger.info("Running the following SalmonTools command: %s",
+                command_str,
+                processor_job=job_context["job_id"])
+
+    completed_command = subprocess.run(command_str.split(),
+                                       stdout=subprocess.PIPE,
+                                       stderr=subprocess.PIPE)
+    job_context['time_end'] = timezone.now()
+
+    # As of SalmonTools 0.1.0, completed_command.returncode is always 0,
+    # (even if error happens).  completed_command.stderr is not totally
+    # reliable either, because it will output the following line even
+    # when the execution succeeds:
+    #  "There were <N> unmapped reads\n"
+    # in which "<N>" is the number of lines in input unmapped_names.txt.
+    #
+    # As a workaround, we are using a regular expression here to test
+    # the status of SalmonTools execution.  Any text in stderr that is
+    # not in the above format is treated as error message.
+    status_str = completed_command.stderr.decode().strip()
+    success_pattern = r'^There were \d+ unmapped reads$'
+    if re.match(success_pattern, status_str):
+        result = ComputationalResult()
+        result.command_executed = command_str
+        result.system_version = __version__
+        result.time_start = job_context['time_start']
+        result.time_end = job_context['time_end']
```

Comment:
If you're just going to use the variables within the same function, you can just use local variables rather than putting these onto the `job_context`. That object is a means for passing data from one step to another.

Reply:
Are you saying that I don't have to set `job_context['time_start']` and `job_context['time_end']` at all?

Reply:
Not exactly. You could just instead use variables named `start_time` and `end_time` rather than putting them in the `job_context`.